Added new generator.

DanielWielanek opened the pull request:
New generator that generate particles by using histogram with Y-Pt distribution.

---

Checklist:

* [ ] Rebased against `dev` branch
* [ ] My name is in the resp. CONTRIBUTORS/AUTHORS file
* [ ] Followed [the seven rules of great commit messages](https://chris.beams.io/posts/git-commit/#seven-rules)

MohammadAlTurany commented:
Hi Daniel,
Could you please add an example that uses this generator to the examples that can also be used for the functional tests, take a look ti the Rutherford example.
best,
Mohammad

DanielWielanek commented:
Hi Mohammad,
do you mean entire example with new detector, digitization etc.? or just adding this as another generator in Rutherford?

MohammadAlTurany commented:
Hi Daniel,
It would be enough if you add it to an existing example, but create a new executable (in ctest) that run it.
best,
Mohammad

dennisklein commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairRoot/pullRequest?prid=5136019&bid=16875393)

dennisklein commented on generators/FairBoxGenerator.cxx:
```diff
@@ -126,26 +114,38 @@ FairBoxGenerator& FairBoxGenerator::operator=(const FairBoxGenerator& rhs)
+    Double_t Y1  = TMath::Min(y1,y2);
+    Double_t Y2 = TMath::Max(y1,y2);
+    Double_t dX = 0.5*(X2 - X1);
+    Double_t dY = 0.5*(Y2 - Y1);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'dY' is assigned a value that is never used.](https://app.codacy.com/manual/FairRootGroup/FairRoot/pullRequest?prid=5136019)

DanielWielanek commented:
Hi Mohammad,
I added this generator to rutherford. I also modified generators, I think that maybe it will be better to create one abstract generator to all "on the fly generators" like FairParticle, FairBox - some parts of them can be "unified" (like generating vertex position, settters/getters for PDG code, multiplicity).
If you think that this is good idea, tomorrow I can made further changes (improvements in log informations).

MohammadAlTurany commented:
Sure it is a good idea, please go ahead.
Thanks. 
Mohammad

DanielWielanek commented:
I removed "Fatal" used in generators (derived from TObject) and replaced by LOG(fatal).

fuhlig1 commented on generators/FairBaseMCGenerator.h:
```diff
@@ -0,0 +1,72 @@
+		kGauss,
+		kExp
+	};
+	FairBaseMCGenerator();
```
I don't see why we need the copy constructor as well as the assignment operator. I don't see the need to copy an existing operator. In my opinion if more than one generator is needed one should create each one separately. If these constructors are really needed please change is to   

FairBaseMCGenerator(const FairBaseMCGenerator) = default;
FairBaseMCGenerator& operator=(const FairBaseMCGenerator) =default;

and delete the implementation. Since all data members are simple types there is no need to implement the constructors explicitly.

DanielWielanek replied:
Hi Florian,
I think not it should be fine. I also remove change constructors for FairBox, this generator also contains only simple types.

fuhlig1 commented on generators/FairYPtGenerator.h:
```diff
@@ -0,0 +1,35 @@
+    Double_t GetMass2()const{return GetPDGMass()*GetPDGMass();};
+    void GetYPt(Double_t &y, Double_t &pt);
+  private:
+    TH2D fYPt;// Y-Pt distribution
```
I think also here no copy constructor is actually needed to be implemented.

karabowi commented:
Dear Daniel,

When running the rutherford example with your pull request in the Multithreaded mode of the Geant4, the simulation crashes at the end with: 
```
root.exe(15568,0x700005088000) malloc: *** error for object 0x7f9e3ce8b4a0: pointer being freed was not allocated
root.exe(15568,0x700005088000) malloc: *** set a breakpoint in malloc_error_break to debug
```
This is triggered by: 
`    TH2D fYPt;// Y-Pt distribution`
Can you test and fix it?
